proposals/extension_get_contexts.md

```diff
@@ -0,0 +1,171 @@
+  SERVICE_WORKER: 'SERVICE_WORKER',         // Service workers for the extension.
+  OFFSCREEN_DOCUMENT: 'OFFSCREEN_DOCUMENT'  // Offscreen documents for the extension.
+};
+```
+
+It would return an array of new `Context` objects for each frame the extension
+is running in:
+
+```
+extension.Context = {
+  tabId: int,                    // ID of the tab for this view.
+  windowId: int,                 // ID of the window this tab is located in.
+  documentId: int,               // ID of the DOM document loaded in the tab.
+  frameId: int,                  // ID of the frame this context is running for.
+  contextType: ContextType,      // Context type -- tab, popup, etc.
+  url: string,                   // The current URL of the tab.
+  origin: string,                // The origin of the context. Mostly this is the
+                                 // extension origin.
+  incognito: boolean             // Whether the context is for an incognito
+                                 // profile.
+}
+```
+
+`Context` objects will contain an `incognito` value indicating if the Context is
+running in an incognito profile. In this area, this brings up the case of an
+extension specifying
+[split incognito mode](https://developer.chrome.com/docs/extensions/mv3/manifest/incognito/#split).
+In this case we would not return the <code>Context</code>s from the related (but
+different) extension process running for the incognito profile.
+<code>spanning</code> mode would return any <code>Context</code>s running in an
+incognito profile (with <code>Context.incognito</code> being <code>true</code>)
+in the same shared process. More comments about this are in the Future Work
+section on this proposal.
+
+## Future Work
+
+### Multi Page Architecture Fields (`DocumentLifeCycle`, `FrameType`, and `parentDocumentId`, and `parentFrameId`)
+
+[Multi Page Architecture](http://doc/1NginQ8k0w3znuwTiJ5qjYmBKgZDekvEPC22q0I4swxQ#heading=h.w1qo2n6sr8wn)
+changed how outermost frames are presented in web navigation in chromium. More
+context on how this can impact extensions is
+[here](https://developer.chrome.com/blog/extension-instantnav/). Considering
+this new architecture it is likely that extensions will want to know what
+navigation state (`DocumentLifeCycle`) a `Context` is in, whether the frame’s
+`FrameType` is an outermost frame, and other fields like `parentDocumentId`, and
+`parentFrameId`. These are fields we would like to include in a future iteration
+of this API.
+
+### Asynchronous and Synchronous Version
+
+A potential avenue is that, since this new API is asynchronous whereas
+`getViews()` is synchronous, we could explore a synchronous version of this new
+API if there’s significant demand for it. As mentioned
+previously<sup>[1](#footnotes)</sup>, there are technical limitations that make
+it infeasible to synchronously fetch all `Context`s. We could provide two
+methods, one synchronous, and one asynchronous API methods:
```

Comment:
`await` is good enough for this. Let's not add any more complex sync APIs.

Reply:
I added a sentence that required the demand to justify why `await` cannot be used. If no others desire this -- I can replace the whole section before commiting mentioning that there is no significant need for this (just for future reference)?